File: backend/websocket_server.py
import json
import time
import threading
from flask import Flask
from flask_socketio import SocketIO
import redis

app = Flask(__name__)
# السماح للواجهة بالاتصال من أي نطاق (CORS)
socketio = SocketIO(app, cors_allowed_origins="*")

# إعداد اتصال Redis
redis_client = redis.Redis(host='localhost', port=6379, decode_responses=True)
STREAM_NAME = 'ndsp.decision.stream'

@socketio.on('connect')
def handle_connect():
    print("🟢 [Frontend Bridge] Client connected")

@socketio.on('disconnect')
def handle_disconnect():
    print("🔴 [Frontend Bridge] Client disconnected")

def broadcast_decisions():
    print(f"📡 [WebSocket Broadcaster] Listening to {STREAM_NAME} to broadcast...")
    last_id = '$' # الاستماع للأحداث الجديدة فقط
    
    while True:
        try:
            # انتظار حدث جديد من Redis
            streams = redis_client.xread({STREAM_NAME: last_id}, count=1, block=5000)
            
            if streams:
                for stream, messages in streams:
                    for message_id, message_data in messages:
                        last_id = message_id
                        payload_str = message_data.get('payload')
                        
                        if payload_str:
                            data = json.loads(payload_str)
                            # بث القرار مباشرة للواجهة
                            socketio.emit('ndsp_final_decision', {
                                'eventId': message_id,
                                'data': data
                            })
                            print(f"➡️ [WebSocket] Broadcasted Decision {message_id} to Frontend.")
        except Exception as e:
            print(f"❌ [WebSocket Error] Failed to read stream: {e}")
            time.sleep(5)

if __name__ == '__main__':
    print("🚀 [API Gateway] WebSocket Server is starting on port 3005...")
    # تشغيل مستمع Redis في مسار (Thread) منفصل لكي لا يوقف الخادم
    threading.Thread(target=broadcast_decisions, daemon=True).start()
    
    # تشغيل خادم WebSockets
    socketio.run(app, host='0.0.0.0', port=3005, debug=False, allow_unsafe_werkzeug=True)

File: backend/src/api/websocketServer.ts
import { Server } from 'socket.io';
import { createServer } from 'http';
import express from 'express';
import cors from 'cors';
import { redis, NDSP_STREAMS } from '../eventbus/redisClient';

const app = express();
app.use(cors());

const httpServer = createServer(app);

// إعداد خادم WebSockets مع السماح بالاتصال من الواجهة
const io = new Server(httpServer, {
cors: {
origin: "*", // في الإنتاج، قم بتحديد رابط الـ Frontend الخاص بك
methods: ["GET", "POST"]
}
});

io.on('connection', (socket) => {
console.log([Frontend Bridge] Client connected: ${socket.id});

socket.on('disconnect', () => {
console.log([Frontend Bridge] Client disconnected: ${socket.id});
});
});

// دالة للاستماع لقرارات NDSP وبثها للواجهة
async function broadcastDecisions() {
let lastId = '$';
console.log([WebSocket Broadcaster] Listening to ${NDSP_STREAMS.DECISION} to broadcast...);

while (true) {
try {
const streamResult = await redis.xread(
'BLOCK', 0,
'STREAMS', NDSP_STREAMS.DECISION, lastId
);

  if (streamResult) {
    const messages = streamResult[0][1];
    for (const message of messages) {
      lastId = message[0];
      const payloadStr = message[1][1]; 
      
      // بث القرار مباشرة لكل المستخدمين المتصلين
      io.emit('ndsp_final_decision', {
        eventId: lastId,
        data: JSON.parse(payloadStr)
      });
      
      console.log(`[WebSocket] Broadcasted Decision ${lastId} to Frontend.`);
    }
  }
} catch (error) {
  console.error('[WebSocket Error] Failed to read stream:', error);
  await new Promise(resolve => setTimeout(resolve, 5000));
}


}
}

const PORT = process.env.PORT || 3001;
httpServer.listen(PORT, () => {
console.log([API Gateway] WebSocket Server is running on port ${PORT});
broadcastDecisions(); // بدء الاستماع والبث
});
